src/Features/Core/Portable/EmbeddedLanguages/Json/JsonParser.cs

```diff
@@ -0,0 +1,572 @@
+                ? ImmutableArray<EmbeddedDiagnostic>.Empty
+                : ImmutableArray.Create(diagnostic.Value));
+
+            static EmbeddedDiagnostic? Earliest(EmbeddedDiagnostic? d1, EmbeddedDiagnostic? d2)
+            {
+                if (d1 == null)
+                    return d2;
+
+                if (d2 == null)
+                    return d1;
+
+                return d1.Value.Span.Start <= d2.Value.Span.Start ? d1 : d2;
+            }
+        }
+
+        /// <summary>
+        /// Checks for errors in json for both json.net and strict mode.
+        /// </summary>
+        private static EmbeddedDiagnostic? CheckTopLevel(
+            VirtualCharSequence text, JsonCompilationUnit compilationUnit)
+        {
+            var sequence = compilationUnit.Sequence;
+            if (sequence.IsEmpty)
+            {
+                // json is not allowed to be just whitespace.
+                if (text.Length > 0 &&
+                    compilationUnit.EndOfFileToken.LeadingTrivia.All(
+                        t => t.Kind is JsonKind.WhitespaceTrivia or JsonKind.EndOfLineTrivia))
+                {
+                    return new EmbeddedDiagnostic(FeaturesResources.Syntax_error, GetSpan(text));
+                }
+
+                return null;
+            }
+            else if (sequence.Length >= 2)
+            {
+                // the top level can't have more than one actual value.
+                var firstToken = GetFirstToken(sequence[1]);
+                return new EmbeddedDiagnostic(
+                    string.Format(FeaturesResources._0_unexpected, firstToken.VirtualChars[0]),
+                    firstToken.GetSpan());
+            }
+            else
+            {
+                var child = sequence.Single();
+
+                // Commas should never show up in the top level sequence.
+                if (child.Kind == JsonKind.CommaValue)
+                {
+                    var emptyValue = (JsonCommaValueNode)child;
+                    return new EmbeddedDiagnostic(
+                        string.Format(FeaturesResources._0_unexpected, ','),
+                        emptyValue.CommaToken.GetSpan());
+                }
+                else if (child.Kind == JsonKind.Property)
+                {
+                    var propertyValue = (JsonPropertyNode)child;
+                    return new EmbeddedDiagnostic(
+                        string.Format(FeaturesResources._0_unexpected, ':'),
+                        propertyValue.ColonToken.GetSpan());
+                }
+
+                return CheckSyntax(child);
+            }
+
+            static EmbeddedDiagnostic? CheckSyntax(JsonNode node)
+            {
+                var diagnostic = node.Kind switch
+                {
+                    JsonKind.Array => CheckArray((JsonArrayNode)node),
+                    _ => null,
+                };
+
+                return diagnostic ?? CheckChildren(node);
+            }
```

Comment:
```suggestion
            static EmbeddedDiagnostic? CheckSyntax(JsonNode node)
            	=> node.Kind switch 
            	{
            		JsonKind.Array => CheckArray((JsonArrayNode)node),
            		_ => CheckChildren(node)
      		};
```

Reply:
that's not quite accurate.  we want CheckChildren when CheckArray returns nothing.

Reply:
oh interesting, that's really subtle. What's the json look like for that scenario?